cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/spark/data/SparkSchemaCreator.java

```diff
@@ -0,0 +1,202 @@
+
+			for (StructField field : resultsSchema.fields()) {
+				fullSchema = fullSchema.add(field);
+			}
+			resultsSchema = fullSchema;
+		}
+		
+		return resultsSchema;
+	}
+
+	private Tuple2<String, DataType> getKeyInformationForContext(String contextName, Set<Tuple2<String, String>> usingInfos) {
+		ContextDefinition contextDefinition = getContextDefinition(contextName);
+		
+		String primaryDataType = contextDefinition.getPrimaryDataType();
+		String primaryKeyColumn = contextDefinition.getPrimaryKeyColumn();
+
+		DataType keyType = null;
+
+		// Check the model info for each non-system using statement from the measures run for this context.
+		// Try to find the key column's type information from a single model info.
+		for (Tuple2<String, String> usingInfo : usingInfos) {
+			VersionedIdentifier modelInfoIdentifier = new VersionedIdentifier().withId(usingInfo._1()).withVersion(usingInfo._2());
+			ModelInfo modelInfo = ModelInfoLoader.getModelInfoProvider(modelInfoIdentifier).load();
+
+			// Look for a ClassInfo element matching primaryDataType for the context
+			List<ClassInfo> classInfos = modelInfo.getTypeInfo().stream()
+					.map(x -> (ClassInfo) x)
+					.filter(x -> x.getName().equals(primaryDataType))
+					.collect(Collectors.toList());
+
+			if (!classInfos.isEmpty()) {
+				if (classInfos.size() == 1) {
+					List<ClassInfoElement> elements = classInfos.get(0).getElement().stream()
+							.filter(x -> x.getName().equals(primaryKeyColumn))
+							.collect(Collectors.toList());
+
+					// A future ModelInfo file may contain the information
+					if (elements.isEmpty()) {
+						continue;
+					}
+					else if (elements.size() == 1) {
+						String elementType = elements.get(0).getElementType();
+
+						// store it
+						if (keyType == null) {
+							keyType = getSparkTypeForSystemValue(elementType);
+						} else {
+							throw new IllegalArgumentException(
+									"Multiple definitions found for " + primaryDataType + "." + primaryKeyColumn
+											+ " in the provided ModelInfo files. Cannot infer key type for context: " + contextName);
+						}
+					}
+					else if (elements.size() > 1) {
+						throw new IllegalArgumentException("ModelInfo " + modelInfoIdentifier + " contains multiple element definitions for " + primaryKeyColumn + " for type " + primaryDataType);
+					}
+				}
+				else {
+					throw new IllegalArgumentException("ModelInfo " + modelInfoIdentifier + " contains multiple definitions for type " + primaryDataType);
+				}
+			}
+
+		}
+		
+		if (keyType == null) {
+			throw new IllegalArgumentException(
+					"Could not locate type information for " + primaryDataType + "." + primaryKeyColumn
+							+ " in the provided ModelInfo files. Cannot infer key type for context: " + contextName);			
+		}
+		return new Tuple2<>(contextDefinition.getPrimaryKeyColumn(), keyType);
+	}
+	
+	private ContextDefinition getContextDefinition(String contextName) {
```

Comment:
This is a good candidate for a helper method `getContextDefinitionByName` on the ContextDefinitions object.

Reply:
Thanks for the suggestion